File: Sources/CountriesFlagsObjC/source/Countries/UIImage+SouthAfrica.m
//
//  UIImage+SouthAfrica.m
//  CountriesFlags
//

#import "Countries/UIImage+SouthAfrica.h"

@implementation UIImage (SouthAfrica)

+ (UIImage * _Nullable)flagSouthAfricaWithSize:(CGSize)size {
    UIColor* colorGreen = [UIColor colorWithRed:0.0f
                                          green:122.0f/255.0f
                                           blue:77.0f/255.0f
                                          alpha:1.0f];
    UIColor* colorBlue = [UIColor colorWithRed:0.0f
                                         green:35.0f/255.0f
                                          blue:149.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:222.0f/255.0f
                                        green:56.0f/255.0f
                                         blue:49.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorGold = [UIColor colorWithRed:1.0f
                                         green:182.0f/255.0f
                                          blue:18.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    UIColor* colorBlack = [UIColor blackColor];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorGreen setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorGold setFill];
    UIBezierPath* path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(0.0f, size.height / 854.0f * 102.0f)];
    [path addLineToPoint:CGPointMake(size.width / 100.0f * 38.0f, size.height / 2.0f)];
    [path addLineToPoint:CGPointMake(0.0f, size.height / 854.0f * 751.0f)];
    [path closePath];
    [path fill];

    [colorBlack setFill];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(0.0f, size.height / 854.0f * 171.0f)];
    [path addLineToPoint:CGPointMake(size.width / 100.0f * 30.0f, size.height / 2.0f)];
    [path addLineToPoint:CGPointMake(0.0f, size.height / 854.0f * 683.0f)];
    [path closePath];
    [path fill];

    [colorWhite setFill];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 1280.0f * 155.0f, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width / 1280.0f * 665.0f, size.height / 854.0f * 341.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height / 854.0f * 341.0f)];
    [path addLineToPoint:CGPointMake(size.width, 0.0f)];
    [path closePath];
    [path fill];

    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 1280.0f * 155.0f, size.height)];
    [path addLineToPoint:CGPointMake(size.width / 1280.0f * 665.0f, size.height / 854.0f * 512.0f)];
    [path addLineToPoint:CGPointMake(size.width,  size.height / 854.0f * 512.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height)];
    [path closePath];
    [path fill];

    [colorRed setFill];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 1280.0f * 259.0f, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width / 1280.0f * 682.0f, size.height / 854.0f * 284.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height / 854.0f * 284.0f)];
    [path addLineToPoint:CGPointMake(size.width, 0.0f)];
    [path closePath];
    [path fill];

    [colorBlue setFill];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 1280.0f * 259.0f, size.height)];
    [path addLineToPoint:CGPointMake(size.width / 1280.0f * 682.0f, size.height / 854.0f * 569.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height / 854.0f * 569.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height)];
    [path closePath];
    [path fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+SouthAfrica.swift
//
//  UIImage+SouthAfrica.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_South_Africa "wikipedia"
     
     Generates the flag of South Africa.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of South Africa.
     */
    internal static func flagSouthAfrica(with size: CGSize) -> UIImage? {
        let colorGreen = UIColor(red: 0.0,
                                 green: 122.0 / 255.0,
                                 blue: 77.0 / 255.0,
                                 alpha: 1.0)
        let colorBlue = UIColor(red: 0.0,
                                green: 35.0 / 255.0,
                                blue: 149.0 / 255.0,
                                alpha: 1.0)
        let colorRed = UIColor(red: 222.0 / 255.0,
                               green: 56.0 / 255.0,
                               blue: 49.0 / 255.0,
                               alpha: 1.0)
        let colorGold = UIColor(red: 1.0,
                                green: 182.0 / 255.0,
                                blue: 18.0 / 255.0,
                                alpha: 1.0)
        let colorWhite = UIColor.white
        let colorBlack = UIColor.black

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorGreen.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorGold.setFill()
        let path = UIBezierPath()
        path.move(to: CGPoint(x: 0.0, y: size.height / 854.0 * 102.0))
        path.addLine(to: CGPoint(x: size.width / 100.0 * 38.0, y: size.height / 2.0))
        path.addLine(to: CGPoint(x: 0.0, y: size.height / 854.0 * 751.0))
        path.close()
        path.fill()

        colorBlack.setFill()
        path.removeAllPoints()
        path.move(to: CGPoint(x: 0.0, y: size.height / 854.0 * 171.0))
        path.addLine(to: CGPoint(x: size.width / 100.0 * 30.0, y: size.height / 2.0))
        path.addLine(to: CGPoint(x: 0.0, y: size.height / 854.0 * 683.0))
        path.close()
        path.fill()

        colorWhite.setFill()
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 1280.0 * 155.0, y: 0.0))
        path.addLine(to: CGPoint(x: size.width / 1280.0 * 665.0, y: size.height / 854.0 * 341.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 854.0 * 341.0))
        path.addLine(to: CGPoint(x: size.width, y: 0.0))
        path.close()
        path.fill()

        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 1280.0 * 155.0, y: size.height))
        path.addLine(to: CGPoint(x: size.width / 1280.0 * 665.0, y: size.height / 854.0 * 512.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 854.0 * 512.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height))
        path.close()
        path.fill()

        colorRed.setFill()
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 1280.0 * 259.0, y: 0.0))
        path.addLine(to: CGPoint(x: size.width / 1280.0 * 682.0, y: size.height / 854.0 * 284.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 854.0 * 284.0))
        path.addLine(to: CGPoint(x: size.width, y: 0.0))
        path.close()
        path.fill()

        colorBlue.setFill()
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 1280.0 * 259.0, y: size.height))
        path.addLine(to: CGPoint(x: size.width / 1280.0 * 682.0, y: size.height / 854.0 * 569.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 854.0 * 569.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height))
        path.close()
        path.fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
